### ISSUE TITLE
LeanDojo hangs while tracing repo

### ISSUE THREAD

**Issue by AG161** [lean-dojo/LeanDojo#17]
**Description**

I created a fork of the Lean 4 example from the docs to test some features of LeanDojo: https://github.com/AG161/lean4-example

The only change in the fork is that I added a few simple theorems to Lean4Example.lean. Then I ran the following:

```
from lean_dojo import LeanGitRepo, trace
repo = LeanGitRepo("https://github.com/AG161/lean4-example", "48825b3")
trace(repo, dst_dir="traced_ast-example")
```

After more than 12 hours the command did not complete, and it didn't look like docker or my CPU was doing much.

**Logs in Debug Mode**

```
(atp) agittis@Andreass-MBP Dojo % export VERBOSE=1
(atp) agittis@Andreass-MBP Dojo % echo $VERBOSE
1
(atp) agittis@Andreass-MBP Dojo % python
Python 3.9.6 (default, May  7 2023, 23:32:44) 
[Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from lean_dojo import LeanGitRepo, trace
>>> repo = LeanGitRepo("https://github.com/AG161/lean4-example", "48825b3")
2023-07-10 22:20:59.495 | DEBUG    | lean_dojo.data_extraction.lean:__post_init__:312 - Querying the commit hash for lean4-example 48825b3
>>> trace(repo, dst_dir="traced_ast-example")
2023-07-10 22:21:31.161 | INFO     | lean_dojo.data_extraction.trace:get_traced_repo_path:132 - Tracing LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-10 22:21:31.163 | DEBUG    | lean_dojo.data_extraction.trace:get_traced_repo_path:134 - Working in the temporary directory /var/folders/wt/sdk361b95ps7x042zc_95zv40000gn/T/tmp_6bjeq8z
2023-07-10 22:21:32.008 | DEBUG    | lean_dojo.data_extraction.lean:clone_and_checkout:370 - Cloning LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-10 22:21:32.422 | DEBUG    | lean_dojo.data_extraction.trace:_trace_lean4:98 - Tracing LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-10 22:21:32.422 | DEBUG    | lean_dojo.container:run:301 - docker run --cidfile 5g452xto.cid --rm -u 501 --mount type=bind,src="/private/var/folders/wt/sdk361b95ps7x042zc_95zv40000gn/T/tmp_6bjeq8z/lean4-example",target="/workspace/lean4-example" --mount type=bind,src="/Users/agittis/ProofSim/atp/lib/python3.9/site-packages/lean_dojo/data_extraction/build_lean4_repo.py",target="/workspace/build_lean4_repo.py" --mount type=bind,src="/Users/agittis/ProofSim/atp/lib/python3.9/site-packages/lean_dojo/data_extraction/ExtractData.lean",target="/workspace/lean4-example/ExtractData.lean" --env NUM_PROCS=8 --workdir /workspace yangky11/lean-dojo python3 build_lean4_repo.py lean4-example
2023-07-11 02:21:33.081 | INFO     | __main__:main:98 - Building lean4-example
info: syncing channel updates for 'nightly'
info: latest update on nightly, lean version nightly-2023-07-06
info: downloading component 'lean'
info: installing component 'lean'
[0/4] Building Lean4Example
[1/4] Compiling Lean4Example
[1/4] Building Main
[2/4] Compiling Main
[4/4] Linking lean4-example
2023-07-11 02:22:54.056 | INFO     | __main__:main:131 - Tracing lean4-example
100%|██████████| 2/2 [00:55<00:00, 27.51s/it]

```

**Platform Information**
-OS:
ProductName: macOS
ProductVersion: 13.4.1
BuildVersion: 22F82

-Python 3.9.6
-LeanDojo Version: 1.1.1
-Pydantic Version: 1.10.10

**Comment by yangky11** [comment]
This might be a bug that has been fixed in the latest version (1.1.2). Please either install it from pypi or the Git repo.

I was unable to reproduce this error. My debug logs:
```
(lean-dojo) kaiyuy@Kaiyus-MacBook-Pro ~ % VERBOSE=1 CONTAINER=native ipython
Python 3.9.16 (main, May 15 2023, 18:51:40) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from lean_dojo import LeanGitRepo, trace
   ...: repo = LeanGitRepo("https://github.com/AG161/lean4-example", "48825b3")
   ...: trace(repo, dst_dir="traced_ast-example")

2023-07-11 18:01:12.385 | DEBUG    | lean_dojo.data_extraction.lean:__post_init__:312 - Querying the commit hash for lean4-example 48825b3
2023-07-11 18:01:13.499 | INFO     | lean_dojo.data_extraction.trace:get_traced_repo_path:132 - Tracing LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-11 18:01:13.502 | DEBUG    | lean_dojo.data_extraction.trace:get_traced_repo_path:134 - Working in the temporary directory /var/folders/kx/4n7dwsnx2gj9sxmyn5h4c93m0000gn/T/tmpj0gowjsh
2023-07-11 18:01:14.493 | DEBUG    | lean_dojo.data_extraction.lean:clone_and_checkout:370 - Cloning LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-11 18:01:15.182 | DEBUG    | lean_dojo.data_extraction.trace:_trace_lean4:98 - Tracing LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-11 18:01:15.227 | DEBUG    | lean_dojo.container:run:183 - NUM_PROCS=8 python3 build_lean4_repo.py lean4-example
2023-07-11 18:01:16.220 | INFO     | __main__:main:102 - Building lean4-example
Building Lean4Example
Compiling Lean4Example
Building Main
Compiling Main
Linking lean4-example
2023-07-11 18:01:25.196 | INFO     | __main__:main:135 - Tracing lean4-example
 86%|███████████████████████████████████▎     | 619/719 [31:21<05:42,  3.43s/it]/Users/kaiyuy/miniconda3/envs/lean-dojo/lib/python3.9/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
2023-07-11 18:33:01.923 | DEBUG    | lean_dojo.data_extraction.traced_data:from_traced_files:1199 - Parsing 620 *.ast.json files in /private/var/folders/kx/4n7dwsnx2gj9sxmyn5h4c93m0000gn/T/tmpj0gowjsh/lean4-example with 7 workers
2023-07-11 18:33:05,375	INFO worker.py:1627 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
100%|█████████████████████████████████████████| 620/620 [00:58<00:00, 10.52it/s]
2023-07-11 18:34:12.568 | DEBUG    | lean_dojo.data_extraction.traced_data:save_to_disk:1246 - Saving 620 traced XML files to /private/var/folders/kx/4n7dwsnx2gj9sxmyn5h4c93m0000gn/T/tmpj0gowjsh/lean4-example with 7 workers
2023-07-11 18:34:14,863	INFO worker.py:1627 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
100%|█████████████████████████████████████████| 620/620 [00:15<00:00, 39.85it/s]
2023-07-11 18:34:42.010 | INFO     | lean_dojo.data_extraction.trace:trace:164 - Loading the traced repo from /Users/kaiyuy/.cache/lean_dojo/AG161-lean4-example-48825b33a1b8a61ad8da20852f53b621f76925d8/lean4-example
2023-07-11 18:34:42.142 | DEBUG    | lean_dojo.data_extraction.traced_data:load_from_disk:1273 - Loading 620 traced XML files from /Users/kaiyuy/.cache/lean_dojo/AG161-lean4-example-48825b33a1b8a61ad8da20852f53b621f76925d8/lean4-example with 7 workers
2023-07-11 18:34:45,495	INFO worker.py:1627 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
100%|█████████████████████████████████████████| 620/620 [01:20<00:00,  7.73it/s]
Out[1]: TracedRepo(repo=LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8'), dependencies={'lean4': LeanGitRepo(url='https://github.com/leanprover/lean4', commit='a3ebfe29ea516e45dba97dfd303cfc11cc47ff04')}, root_dir=PosixPath('/Users/kaiyuy/.cache/lean_dojo/AG161-lean4-example-48825b33a1b8a61ad8da20852f53b621f76925d8/lean4-example'))
```

**Comment by AG161** [comment]
I've upgraded to LeanDojo 1.1.2 and now I get the following output:

```
(atp) agittis@Andreass-MBP Dojo % export VERBOSE=1
(atp) agittis@Andreass-MBP Dojo % echo $VERBOSE
1
(atp) agittis@Andreass-MBP Dojo % python
Python 3.9.6 (default, May  7 2023, 23:32:44) 
[Clang 14.0.3 (clang-1403.0.22.14.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from lean_dojo import LeanGitRepo, trace
>>> repo = LeanGitRepo("https://github.com/AG161/lean4-example", "48825b3")
2023-07-11 21:08:22.926 | DEBUG    | lean_dojo.data_extraction.lean:__post_init__:312 - Querying the commit hash for lean4-example 48825b3
>>> trace(repo, dst_dir="traced_ast-example")
2023-07-11 21:08:28.174 | INFO     | lean_dojo.data_extraction.trace:get_traced_repo_path:132 - Tracing LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-11 21:08:28.175 | DEBUG    | lean_dojo.data_extraction.trace:get_traced_repo_path:134 - Working in the temporary directory /var/folders/wt/sdk361b95ps7x042zc_95zv40000gn/T/tmpwetky6at
2023-07-11 21:08:29.124 | DEBUG    | lean_dojo.data_extraction.lean:clone_and_checkout:370 - Cloning LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-11 21:08:29.503 | DEBUG    | lean_dojo.data_extraction.trace:_trace_lean4:98 - Tracing LeanGitRepo(url='https://github.com/AG161/lean4-example', commit='48825b33a1b8a61ad8da20852f53b621f76925d8')
2023-07-11 21:08:29.504 | DEBUG    | lean_dojo.container:run:305 - docker run --cidfile z7rckje1.cid --rm -u 501 --mount type=bind,src="/private/var/folders/wt/sdk361b95ps7x042zc_95zv40000gn/T/tmpwetky6at/lean4-example",target="/workspace/lean4-example" --mount type=bind,src="/Users/agittis/ProofSim/atp/lib/python3.9/site-packages/lean_dojo/data_extraction/build_lean4_repo.py",target="/workspace/build_lean4_repo.py" --mount type=bind,src="/Users/agittis/ProofSim/atp/lib/python3.9/site-packages/lean_dojo/data_extraction/ExtractData.lean",target="/workspace/lean4-example/ExtractData.lean" --env NUM_PROCS=8 --workdir /workspace yangky11/lean-dojo python3 build_lean4_repo.py lean4-example
2023-07-12 01:08:30.190 | INFO     | __main__:main:102 - Building lean4-example
info: syncing channel updates for 'nightly'
info: latest update on nightly, lean version nightly-2023-07-11
info: downloading component 'lean'
info: installing component 'lean'
[0/4] Building Lean4Example
[1/4] Compiling Lean4Example
[1/4] Building Main
[2/4] Compiling Main
[4/4] Linking lean4-example
2023-07-12 01:10:41.256 | INFO     | __main__:main:135 - Tracing lean4-example
  0%|          | 0/718 [00:20<?, ?it/s]
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib64/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib64/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/workspace/build_lean4_repo.py", line 68, in _monitor
    list(
  File "/usr/lib64/python3.11/pathlib.py", line 953, in glob
    for p in selector.select_from(self):
  File "/usr/lib64/python3.11/pathlib.py", line 408, in _select_from
    for p in successor_select(starting_point, is_dir, exists, scandir):
  File "/usr/lib64/python3.11/pathlib.py", line 355, in _select_from
    with scandir(parent_path) as scandir_it:
         ^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.11/pathlib.py", line 938, in _scandir
    return os.scandir(self)
           ^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'lake-packages/lean4/lib/lean/Lake'


```

Also after the error, the command doesn't halt and return control, it just hangs. I see it saying some stuff about python 3.11 but I am using python 3.9.6.

**Comment by yangky11** [comment]
Unfortunately I wasn't able to reproduce the error.

The `python 3.11` is installed inside docker and is expected.

The problem was that the progress bar didn't work (probably related to [this weird error](https://stackoverflow.com/questions/71787077/python-pathlib-finds-a-file-but-then-looses-it)).

Could you please try the [dev](https://github.com/lean-dojo/LeanDojo/tree/dev) branch? 

Alternatively, you can just ignore the error (since it's just a problem with the progress bar). The tracing was probably working (not hanging). You can`cd` to the temporary directory printed in debug mode (`/var/folders/wt/sdk361b95ps7x042zc_95zv40000gn/T/tmpwetky6at` in the log your pasted) and check the number of generated `*.ast.json` files by something like `ls -lR lean4-example | grep `\.ast\.json``.

**Comment by AG161** [comment]
Hi @yangky11, you were absolutely right - when I navigated to the temporary directory and used your command I could see that the number of .ast.json files were increasing. I also made a separate venv with the dev branch and the progress bar worked! No other errors showed up either. It will take some number of hours before the trace actually completes, but I think the issue is fixed. Thanks for your help!
